## TPM

**TPM**<br>

**Transcripts Per Million (TPM)** is a normalization method for RNA-seq, should be read as "for every 1,000,000 RNA molecules in the RNA-seq sample, x(Normalized Value) came from this gene/transcript."

**STEPS**<br>
**1. Normalize by Gene Length**<br>
For each transcript in the gene model, the number (raw count) of reads mapped is divided by the transcript's length, giving a normalized transcript-level expression.<br>
**2. Determine the Scaling Factor**<br>
The sum of ALL normalized transcript expression values is divided by 1,000,000, to create a scaling factor.<br>
**3. Normalize for Sequencing Depth**<br>
Each transcript's normalized expression is divided by the scaling factor, which results in the TPM value.<br>
**4. Check: Gene-level TPM's**<br>
Gene-level TPM's are calculated by summing up the transcript-level TPM for each gene.
In this scaling, the sum of all TPMs (transcript-level or gene-level) should always equal 1,000,000. 

**Demo:**<br>

Read_Count = <br>

<table style="width:50%">
  <tr> 
    <th>Gene</th>
    <th>T1</th> 
    <th>T2</th>
    <th>T3</th>
    <th>Length</th>
  </tr>
  <tr>
    <td>A</td>
    <td>20</td>
    <td>20</td>
    <th>20</th>
    <th>2</th>
  </tr>
  <tr>
    <td>B</td>
    <td>15</td>
    <td>9</td>
    <th>30</th>
    <th>3</th>
  </tr>
  <tr>
    <td>C</td>
    <td>20</td>
    <td>10</td>
    <th>30</th>
    <th>5</th>
  </tr>
</table>

1. Normalize by gene length <br>
Divide each read count by gene length <br>

<table style="width:50%">
  <tr> 
    <th>Gene</th>
    <th>T1</th> 
    <th>T2</th>
    <th>T3</th>
  </tr>
  <tr>
    <td>A</td>
    <td>10</td>
    <td>10</td>
    <th>10</th>
  </tr>
  <tr>
    <td>B</td>
    <td>5</td>
    <td>3</td>
    <th>10</th>
  </tr>
  <tr>
    <td>C</td>
    <td>4</td>
    <td>2</td>
    <th>6</th>
  </tr>
</table>

2. Determine the Scaling Factor <br>
sum of ALL normalized transcript expression values is divided by 1,000,000 <br>

<table style="width:50%">
  <tr> 
    <th></th>
    <th>T1</th> 
    <th>T2</th>
    <th>T3</th>
  </tr>
  <tr>
    <td>Sum</td>
    <td>19</td>
    <td>15</td>
    <th>26</th>
  </tr>
  <tr>
    <td>SF(x10^-5)</td>
    <td>1.9</td>
    <td>1.5</td>
    <th>2.6</th>
  </tr>
</table>

3. Normalize for sequencing depth <br>
Divide each by normalized transcript expression values the scaling factor to get the TPM <br>

<p style="color:#FF0000";><b>TPM = <b></p> 


<table style="width:50%">
  <tr> 
    <th>Gene</th>
    <th>T1</th> 
    <th>T2</th>
    <th>T3</th>
  </tr>
  <tr>
    <td>A(x10^5) </td>
    <td>5.30</td>
    <td>2.60</td>
    <th>2.10</th>
  </tr>
  <tr>
    <td>B(x10^5) </td>
    <td>6.70</td>
    <td>2.00</td>
    <th>1.30</th>
  </tr>
  <tr>
    <td>C(x10^5) </td>
    <td>3.85</td>
    <td>3.85</td>
    <th>2.30</th>
  </tr>
</table>

4. Check: Gene-level TPM's

<table style="width:50%">
  <tr> 
    <th></th>
    <th>T1</th> 
    <th>T2</th>
    <th>T3</th>
  </tr>
  <tr>
    <td>Sum(x10^5)</td>
    <td>10</td>
    <td>10</td>
    <th>10</th>
  </tr>
</table>

Each Transcript sums up to 1,000,000 showing that TPM was properly calculated

## Master_Count Data to TPM

#### IMPORT LIBRARIES

In [1]:
import pandas as pd
import numpy as np
import csv

#### IMPORT MASTER_COUNT DATA

This is the data to be normalized<br>
The workstation path for this data: 

In [2]:
data = pd.read_csv(r'C:\Users\User\Desktop\new tpm\master_counts_table.tsv')

In [3]:
data.head()

feature  10934H  20729M  11620P  11126H  12534B  18332T  16425Q  \
0  ENSG00000223972       0       0       0       0       0       0       0   
1  ENSG00000227232       0      14       3       1       1       2       6   
2  ENSG00000278267       0       3       1       0       0       0       0   
3  ENSG00000243485       0       0       0       0       0       0       0   
4  ENSG00000274890       0       0       0       0       0       0       0   

   10088U  14557X  ...  12257B  19461J  16606U  18381G  20939X  20352R  \
0       0       0  ...       0       0       0       1       1       0   
1       0       6  ...       8      12      10       5      13      21   
2       0       0  ...       2       2       7       3       2       5   
3       0       0  ...       0       0       0       0       0       0   
4       0       0  ...       0       0       0       0       0       0   

   20956X  19954G  20848U  12765U  
0       1       0       0       1  
1       9       8      16      11  
2       2       5       4       3  
3       0       0       0       0  
4       0       0       0       0  

[5 rows x 2656 columns]

#### IMPORT GENE LENGTH <br>
The Gene Length was generated using the gencode.v29.primary_assembly.annotation.gtf

**Method**: longest_isoform merged <br>
**Workstation Path to script**:/data/ccsb/lucy/newtpm/code; Run: code_lengths.py<br>
**Workstation path to generated Gene Length**: /data/ccsb/lucy/newtpm/code/ncbi_ensembl_coding_mergedgenelength.csv.gz<br>

**Reference**:
1. https://www.biostars.org/p/307603/ <br>
2. https://gist.github.com/slowkow/8101509 <br>
3. https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5199132/<br>
4. http://genomespot.blogspot.com/2019/01/using-gtf-tools-to-get-gene-lengths.html

In [4]:
Gene_Length = pd.read_csv(r'C:\Users\User\Desktop\new tpm\ncbi_ensembl_coding_mergedgenelength.csv', delimiter = '\t')

In [5]:
Gene_Length.head()

Ensembl_gene_identifier  length
0         ENSG00000000003    4535
1         ENSG00000000005    1610
2         ENSG00000000419    1207
3         ENSG00000000457    6883
4         ENSG00000000460    5967

#### MERGE THE GENE LENGTH AND THE MASTER_COUNT DATA

This action appends the gene length as the last column on the Master_Count data.<br> The resulting dataset has 56972 rows × 2656 columns due to missing genes/exons 

In [6]:
mergedDf = data.merge(Gene_Length, left_on='feature', right_on='Ensembl_gene_identifier')
del mergedDf['Ensembl_gene_identifier']
mergedDf.head()

feature  10934H  20729M  11620P  11126H  12534B  18332T  16425Q  \
0  ENSG00000223972       0       0       0       0       0       0       0   
1  ENSG00000227232       0      14       3       1       1       2       6   
2  ENSG00000278267       0       3       1       0       0       0       0   
3  ENSG00000243485       0       0       0       0       0       0       0   
4  ENSG00000237613       0       0       0       0       0       0       0   

   10088U  14557X  ...  19461J  16606U  18381G  20939X  20352R  20956X  \
0       0       0  ...       0       0       1       1       0       1   
1       0       6  ...      12      10       5      13      21       9   
2       0       0  ...       2       7       3       2       5       2   
3       0       0  ...       0       0       0       0       0       0   
4       0       0  ...       0       0       0       0       0       0   

   19954G  20848U  12765U  length  
0       0       0       1    1735  
1       8      16      11    1351  
2       5       4       3      68  
3       0       0       0    1021  
4       0       0       0    1219  

[5 rows x 2657 columns]

In [ ]:
# Save the new dataset as a CSV if needed
#mergedDf.to_csv('C:\\Users\\User\\Desktop\\new tpm\code\Master_Count_With Length.csv')

In [7]:
#Save the sample/row names to a list
Columns = list(mergedDf.columns) 
Columns.remove('feature')
Columns.remove('length')

#Save the feature/column names to a list
feature = list(mergedDf['feature'])

In [8]:
#Set the samle names as index
newmergedDf = mergedDf.set_index('feature')
newmergedDf.head()

10934H  20729M  11620P  11126H  12534B  18332T  16425Q  \
feature                                                                   
ENSG00000223972       0       0       0       0       0       0       0   
ENSG00000227232       0      14       3       1       1       2       6   
ENSG00000278267       0       3       1       0       0       0       0   
ENSG00000243485       0       0       0       0       0       0       0   
ENSG00000237613       0       0       0       0       0       0       0   

                 10088U  14557X  22389Y  ...  19461J  16606U  18381G  20939X  \
feature                                  ...                                   
ENSG00000223972       0       0       0  ...       0       0       1       1   
ENSG00000227232       0       6       0  ...      12      10       5      13   
ENSG00000278267       0       0       1  ...       2       7       3       2   
ENSG00000243485       0       0       0  ...       0       0       0       0   
ENSG00000237613       0       0       0  ...       0       0       0       0   

                 20352R  20956X  19954G  20848U  12765U  length  
feature                                                          
ENSG00000223972       0       1       0       0       1    1735  
ENSG00000227232      21       9       8      16      11    1351  
ENSG00000278267       5       2       5       4       3      68  
ENSG00000243485       0       0       0       0       0    1021  
ENSG00000237613       0       0       0       0       0    1219  

[5 rows x 2656 columns]

#### CREATE A FUNCTION FOR TPM NORMALIZATION

In [9]:
def read_counts2tpm(df):
    """
    convert read counts to TPM (transcripts per million)
    :df: a dataFrame that contains the read count with its gene length. 
    :sample_reads: read count values for all transcripts
    :gene_len: Gene length values
    :return: TPM
    """
    result = df
    sample_reads = result.loc[:, result.columns != 'length'].copy()
    gene_len = result.loc[:, ['length']]
    normalize_by_genelength = sample_reads.values / gene_len.values
    scaling_factor = (np.sum(normalize_by_genelength, axis=0).reshape(1, -1))/1e6
    normalize_sequencingdepth = normalize_by_genelength / scaling_factor
    tpm = normalize_sequencingdepth
    return tpm

#### TPM DATA

TPM_MasterCount = pd.DataFrame(read_counts2tpm(newmergedDf))

In [11]:
#Append the row/column names to the normalized data
TPM_MasterCount.columns = Columns
TPM_MasterCount.insert(0, 'feature', feature, True)
TPM_MasterCount.set_index('feature')
TPM_MasterCount.head()

feature  10934H     20729M    11620P    11126H    12534B   18332T  \
0  ENSG00000223972     0.0   0.000000  0.000000  0.000000  0.000000  0.00000   
1  ENSG00000227232     0.0   2.670914  0.657628  0.170653  0.233875  0.56986   
2  ENSG00000278267     0.0  11.371024  4.355171  0.000000  0.000000  0.00000   
3  ENSG00000243485     0.0   0.000000  0.000000  0.000000  0.000000  0.00000   
4  ENSG00000237613     0.0   0.000000  0.000000  0.000000  0.000000  0.00000   

     16425Q  10088U    14557X  ...    12257B    19461J     16606U    18381G  \
0  0.000000     0.0  0.000000  ...  0.000000  0.000000   0.000000  0.098759   
1  1.653449     0.0  1.266512  ...  1.706881  1.697444   1.456574  0.634149   
2  0.000000     0.0  0.000000  ...  8.477930  5.620704  20.257087  7.559425   
3  0.000000     0.0  0.000000  ...  0.000000  0.000000   0.000000  0.000000   
4  0.000000     0.0  0.000000  ...  0.000000  0.000000   0.000000  0.000000   

     20939X     20352R    20956X     19954G     20848U     12765U  
0  0.134449   0.000000  0.119829   0.000000   0.000000   0.252014  
1  2.244639   4.030325  1.384999   1.589075   2.685463   3.560095  
2  6.860875  19.065018  6.114815  19.731985  13.338459  19.290193  
3  0.000000   0.000000  0.000000   0.000000   0.000000   0.000000  
4  0.000000   0.000000  0.000000   0.000000   0.000000   0.000000  

[5 rows x 2656 columns]

#### Check<br>

The sum of TPM normalized values accross each transcript is 1,000,000 showing that the data was properly normalized. 

In [12]:
TPM_MasterCount.sum(axis = 0, skipna = 
                    True)

feature    ENSG00000223972ENSG00000227232ENSG00000278267E...
10934H                                                 1e+06
20729M                                                 1e+06
11620P                                                 1e+06
11126H                                                 1e+06
                                 ...                        
20352R                                                 1e+06
20956X                                                 1e+06
19954G                                                 1e+06
20848U                                                 1e+06
12765U                                                 1e+06
Length: 2656, dtype: object

#### Save the TPM_Normalized data as a CSV file

In [ ]:
#TPM_MasterCount.to_csv('C:\\Users\\User\\Desktop\\new tpm\code\Master_Count_TPM.csv')